### PR TITLE
Work on Harold's `Maybe` `bind` example

### PR DESCRIPTION
**Pull request by mark-moir**
This PR starts with an example created by @haroldcarr, with a simple `MaybeD` (should be renamed `MaybeAST`) program that includes two `bind`s using (unknown) `MaybeD` programs.  It then finishes the proof that came with the initial example, and along the way establishes a simpler example with only a single bind and proves it in a roundabout way, with documentation to help us  understand it.  It then postulates a lemma to make it easier to provide the proof obligations for `bind`s in `MaybeD`, and uses that lemma to also prove the two-bind example.  Finally, it begins the proof of this lemma, which is not quite finished as I am out of time.  The above-mentioned lemma is not quite satisfying because it requires explicitly specifying the "continuation" after the bind, which is cumbersome and takes us back to the "breaking into steps" days we want to leave behind in order to avoid unsolved constraints warnings, so this requires some more work too.

### CODE REVIEW

File: src/Dijkstra/AST/Maybe.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Maybe.agda:8]: Since `obm-dangerous-magic` is being used, I do think this is necessary.

File: src/Dijkstra/AST/Maybe.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Maybe.agda:22]: why?

File: src/Dijkstra/AST/Maybe.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Maybe.agda:191]: approved modulo finishing these TODOs.

**Review by haroldcarr — APPROVED** [review]
approved modulo `obm-dangerous-magic`

File: src/Dijkstra/AST/Maybe.agda
Comment (mark-moir) [reply src/Dijkstra/AST/Maybe.agda:8]: Yes, but TEMPORARY!!

File: src/Dijkstra/AST/Maybe.agda
Comment (mark-moir) [reply src/Dijkstra/AST/Maybe.agda:22]: No good reason.  Not intentional.

File: src/Dijkstra/AST/Maybe.agda
Comment (mark-moir) [reply src/Dijkstra/AST/Maybe.agda:191]: Thanks.  I'll keep working on eliminating the dangerous magic before merging and will request re-review when done.

**Comment by mark-moir** [comment]
@haroldcarr I've finished this off and update the description, would appreciate a quick look before I merge.

File: src/Dijkstra/AST/Either.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Either.agda:39]: I notice you changed `MaybeSyntax` and `RWSSyntax` also.
For consistency, in a new PR, we should update `RWS` to `RWSAST` and `runRWSAST` etc

File: src/Dijkstra/AST/Examples/Maybe/Bind.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Examples/Maybe/Bind.agda:5]: curious : have you tried deleting the rest of the imports one-by-one to see which are really needed?

File: src/Dijkstra/AST/Examples/Maybe/Bind.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Examples/Maybe/Bind.agda:40]: perhaps add a comment about what this comment means

File: src/Dijkstra/AST/Examples/Maybe/Partiality.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Examples/Maybe/Partiality.agda:106]: this change is nice - easier on the eyes

File: src/Dijkstra/AST/Maybe.agda
Comment (haroldcarr) [inline src/Dijkstra/AST/Maybe.agda:7]: Modules under `Dijkstra` should not import `LibraBFT` modules.

Whatever is being used from `Libra.Base.Util` should be moved to 
- `Haskell.*` : if it is a "haskell" thing
- `Util.Prelude` :  otherwise

**Review by haroldcarr — APPROVED** [review]
modulo NOT importing `LibraBFT`

File: src/Dijkstra/AST/Either.agda
Comment (mark-moir) [reply src/Dijkstra/AST/Either.agda:39]: Will do.

File: src/Dijkstra/AST/Examples/Maybe/Bind.agda
Comment (mark-moir) [reply src/Dijkstra/AST/Examples/Maybe/Bind.agda:5]: This issue will be superseded by another PR I plan on doing soon, which streamlines the user experience for the framework.

File: src/Dijkstra/AST/Examples/Maybe/Bind.agda
Comment (mark-moir) [reply src/Dijkstra/AST/Examples/Maybe/Bind.agda:40]: DONE

File: src/Dijkstra/AST/Maybe.agda
Comment (mark-moir) [reply src/Dijkstra/AST/Maybe.agda:7]: Thanks, this was just for `obm-dangerous-magic` and was temporary.  DONE.

**Pull request merged**
